enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,232 @@
+
+#### Story 4:
+
+We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain, and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.
+
+
+### Implementation Details
+
+The `external` `cloud-controller-manager` implementation will be hosted within openshift repository in https://github.com/openshift/cluster-cloud-controller-manager-operator. That repository will manage `CCM` provisioning for metioned cloud providers + `Openstack` out-of-tree provider implementation described in the [document](https://github.com/openshift/enhancements/tree/master/enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md)
+
+`Cloud-controller` will be provisioned in the `openshift-cloud-controller-manager` namespace.
+
+Preparation steps, common for all providers:
+
+- Add support for recognition of the `external` option in [library-go](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) which will disable config parsing and add support for `external` field in the `Infrastructure` resource.
+- Port the templates of `cloud-controller` from upstream repositories into proposed https://github.com/openshift/cluster-cloud-controller-manager-operator repository. Those templates will belong to the newly created `openshift-cloud-controller-manager` namespace and be managed with an operator.
+- Operator will be deployed in the `openshift-cloud-controller-manager-operator` namespace.
+
+Each provider implementation will need to do the following minimal set of actions in order to make out-of-tree implemetation work:
+
+- Use `external` flag in the `kube-controller-manager` pod instead of the previously used provider. Remove the `cloud-config` option from the template.
+- Use `external` flag in the `kubelet` pod. Remove the `cloud-config` option from the template.
+- Provision a `cloud-controller-manager` `DaemonSet` in the `openshift-cloud-controller-manager` namespace.
+
+#### API resource changes
+
+In order to achieve desired cluster configuration with `external` cloud components, we need to add a new field into `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11). With the addition, a `status.platformStatus.external` field will determine the architectural change in the cluster configuration. Other components will be able to consume the configuration same way as before, only with a minor change of the resource processing.
```

Comment:
who set this value and when?

Reply:
It is populated by the installer for every platform: https://github.com/openshift/installer/blob/0f625f6cc1028fd0a947dbee6c53b200c47466bf/pkg/asset/manifests/infrastructure.go#L56

Reply:
infrastructure.status.platformStatus.external is a very generic name.

Reply:
Why do we need an API field for this? Surely this is internal to the state of the cluster and we should be able to handle this without exposing it as an API field to the customer?

I had imagined everything would be internal or external based on the `platform` given a particular version of OCP, there wouldn't be a choice or logic within any controller to determine internal or external.

Where KCM and KA are configured, we need to return empty string and have the `--cloud-provider` flag ommitted, there's a switch that builds this already which could just be updated, I imagine as we upgrade to OCP n+1 there is a component that can reconfigure these?

Where `Kubelet` is configured, we need to set `--cloud-provider=external` based on a list of providers being external or not, I imagine there's some switch on `platform` already that determines the value for this flag, we just need to update that. Will MCO be able to update that on each node during an upgrade?

The operator or whatever deploys CCM will be able to look at the `platform`, and deploy the correct binary dependent on that.